src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -48,5 +50,129 @@ public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private Dictionary<Tuple<string, string>, EventBase> EventsInProgress = new Dictionary<Tuple<string, string>, EventBase>();
```

Comment:
The dictionary needs to be protected against access by multiple thread. Can you please do that.